design/storage-class-device-set.md

```diff
@@ -0,0 +1,69 @@
+    Name                 string // A unique identifier for the set
+    Count                int // Number of devices in this set
+    Resources            v1.ResourceRequirements // Requests/limits for the devices
+    Placement            rook.Placement // Placement constraints for the devices
+    Config               map[string]string // Provider-specific device configuration
+    volumeClaimTemplates []v1.PersistentVolumeClaimSpec // List of PVC templates for the underlying storage devices
+}
+```
+
+A provider will be able to use the StorageClassDeviceSet struct to describe the properties of a particular set of StorageClassDevices. In this design, the notion of a "StorageClassDevice" is an abstract concept, separate from underlying storage devices. In particular, a single StorageClassDevice could be comprised of multiple underlying storage devices, specified by having more than one item in the volumeClaimTemplates field.
+
+A StorageClassDeviceSet will have the following fields associated with it:
+
+* **name**: A name for the set. **[required]**
+* **count**: The number of devices in the set. **[required]**
+* **resources**: The CPU and RAM requests/limits for the devices. Default is no resource requests.
+* **placement**: The placement criteria for the devices. Default is no placement criteria.
+* **config**: Granular device configuration. This is a generic `map[string]string` to allow for provider-specific configuration.
+* **volumeClaimTemplates**: A list of PVC templates to use for provisioning the underlying storage devices.
+
+  A volumeClaimTemplate must specify the following fields:
+  *  **resources.requests.storage**: The desired capacity for the underlying storage devices.
+  *  **storageClassName**: The StorageClass to provision PVCs from.
+
+## Example Workflow: rook-ceph OSDs
+The CephCluster CRD could be extended to include a new field: `spec.storage.StorageClassDeviceSets`, which would be a list of one or more StorageClassDeviceSets. If elements exist in this list, the CephCluster controller would then create enough PVCs to match each StorageClassDeviceSet's `Count` field, attach them to individual OsdPrepare Jobs, then attach them to OSD Pods once the Jobs are completed. For the initial implementation, only one volumeClaimTemplate would be supporter, if only to tighten the scope for an MVP.
+
+The PVCs would be provisioned against a configured or default StorageClass. It is recommended that the admin setup a StorageClass with `volumeBindingMode: WaitForFirstConsumer` set.
+
+If the admin wishes to control device placement, it will be up to them to make sure the desired nodes are labeled properly to ensure the Kubernetes scheduler will distribute the OSD Pods based on Placement criteria.
+
+### OSDs as StatefulSets
+A set of OSDs could be represented by a StatefulSet. This would allow rook-ceph to offload the entire notion of Pod and PVC creation and management to Kubernetes while still maintaining OSD identity.
```

Comment:
```suggestion
A set of OSDs could be represented by a StatefulSet. This would allow Rook-Ceph to offload the entire notion of Pod and PVC creation and management to Kubernetes while still maintaining OSD identity.
```